### ISSUE TITLE
pyenv incorrectly labels brew python as system python

### ISSUE THREAD

**Issue by caleb15** [pyenv/pyenv#1436]
### Prerequisite
* [X] Make sure your problem is not listed in [the common build problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
* [X] Make sure no duplicated issue has already been reported in [the pyenv issues](https://github.com/pyenv/pyenv/issues). You should look for closed issues, too.
* [X] Make sure you are not asking us to help solving your specific issue.
  * GitHub issues is opened mainly for development purposes. If you want to ask someone to help solving your problem, go to some community site like [Gitter](https://gitter.im/yyuu/pyenv), [StackOverflow](https://stackoverflow.com/questions/tagged/pyenv), etc.
* [?] Make sure your problem is not derived from packaging (e.g. [Homebrew](https://brew.sh)).
  * Please refer to the package documentation for the installation issues, etc.
* [X] Make sure your problem is not derived from plugins.
  * This repository is maintaining `pyenv` and the default `python-build` plugin only. Please refrain from reporting issues of other plugins here.

### Description
- [ubuntu 18] Platform information (e.g. Ubuntu Linux 16.04):
- [N/A] OS architecture (e.g. amd64):
- [ 1.2.13] pyenv version:

1. install brew
2. install python via brew, or a package that requires python. For example, `brew install cheat`
3. install pyenv
4. run `pyenv versions`
5. pyenv reports a "system" python and  "3.6.4" python. The 3.6.4 python is the system python that comes with ubuntu. The "system" python is actually the homebrew python
6. Set pyenv to use system python if it's not already
7. `python -V` reports 3.7, the version of the homebrew python. The system python should be `3.6.4`

Oddly when I uninstalled the brew python the system python switched over to `3.6.8`. Not sure why it chose that as the system python.

**Comment by caleb15** [comment]
@wobo-aiguofer says:

> Btw, the reason pyenv thinks your system python is the one from brew is due to your $PATH. Whichever python it finds first in PATH is the one that will be used for system.

That explains this behavior. I double-checked the readme and it's (sorta) mentioned there as well:

>  pyenv assumes you want to use the "system" Python. (In other words, whatever version would run if pyenv weren't in your PATH.)

"system" is a bit of a misnomer, because the first python found in the PATH may or may not actually be the system python. Maybe a better name would be "default" python?

What would also be helpful is if the version was displayed next to the name. For example, currently `pyenv versions` displays this o my machine:

```
* system (set by /home/caleb/.pyenv/version)
  3.6.8
  3.6.8/envs/deploy
  deploy
```

but it would be helpful if it displayed this:

```
* system (set by /home/caleb/.pyenv/version) version:3.7.6
  3.6.8
  3.6.8/envs/deploy
  deploy version:3.6.8
```

**Comment by caleb15** [comment]
Another option is to not bother giving the first python found in PATH a name - just have it be one of the versions available. This would simplify things.

**Comment by aiguofer** [comment]
Hmm, it would be pretty cool if it could display every version in the `$PATH` as a separate version. 

One way to achieve this and maintain a "pyenv way" of doing things could be to add symlinks in `versions`.. although that could get ugly when `$PATH` changes...

The idea of adding the version to each name could be pretty cool too. If I have some time I might take a stab at a PR.

**Comment by joshfriend** [comment]
please note that the behavior you are describing would also apply to rbenv and should probably be applied there first so we can merge it in to pyenv as well

**Comment by native-api** [comment]
Pyenv doesn't try to do any of that because "system" Pythons are outside its control. We basically can't rely on anything about them and their behavior.

If we try to emulate them as "real" Pyenv Pythons, we'll set up out users up to fall prey to [The Law of Leaky Abstractions](https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/) -- any functionality we add would be prone to breaking, breaking other things or giving unexpected results.

E.g.:

* the version can change at any time
* and version number alone is not sufficient 'cuz there are many flavors of Python and many possible compile-time configurations
  * note that "version numbers" that `pyenv versions` give you are not "real" versions but simply names under which those Pythons installations are placed
* any executable we would rely on can be missing, belong to another installation or be an unrelated piece of software altogether
  * and we cannot do anything about it (and even tell which is the case!) 'cuz they are outside of our control -- who knows if this is normal for them and/or the user specifically set it up like that?
* arbitrary, system-specifc additional steps can be needed to use any of their functionality
  * see e.g. https://github.com/pyenv/pyenv/issues/1204#issuecomment-834916966 where a relogin was needed to use scripts installed by Pip -- _but only the first time around!_

**Comment by caleb15** [comment]
Sorry, I should have created a new issue for the version suggestion. The version suggestion is a seperate enhancement issue, the main issue here is a that the "system" name is misleading. @native-api want me to create a new ticket for the versions and we can re-label this as a bug?

**Comment by native-api** [comment]
Create it if you want, but as per above, it's not of much use and can have bad side effects -- so the above concerns need to be addressed somehow if you hope to get anywhere with it.

**Comment by native-api** [comment]
E.g. you can register an existing external Python installation with Pyenv as per https://github.com/pyenv/pyenv/issues/1244, under a descriptive name that would be telling you everything you want to be reminded of about it.

**Comment by caleb15** [comment]
Eh, I took a look at it again and I'm not sure if showing the version would actually be that useful.

Can you re-open this issue please? This issue is for the path python being labeled incorrectly, not the version suggestion.

**Comment by native-api** [comment]
With the initial message's steps in MacOS 11 and Pyenv latest head, `pyenv versions` shows just "`system`", as expected.

**Comment by caleb15** [comment]
The issue is that `system` is misleading, because it may or not not actually be the system python. What do you think about `default` instead?

**Comment by native-api** [comment]
`system` means "out there", "somewhere in the larger system" -- ie. outside Pyenv's environment.

`default` would be just as misleading since `pyenv global` is specifically designed to override that "default".

**Comment by native-api** [comment]
`system` means "out there", "somewhere in the larger system" -- i.e. outside Pyenv's environment.

`default` would be even more misleading since `pyenv global` is specifically designed to override that "default".

**Comment by native-api** [comment]
"`default`" is also confusing because it doesn't tell anything about the nature of that "default" and where it comes from:

```
$ pyenv versions
default
3.7.8
3.9.5
pypy3-2.4.0
```
What's this "default" thing? Is it a separate installation? Or is it equal to one of the other ones -- e.g. the one that is currently selected? Would it change somehow from some of the commands?

See? It leads the user into a completely wrong train fo thought.

**Comment by caleb15** [comment]
`system` could also be taken to mean the python bundled with the system, `/usr/bin/python3`. Good point about `default` :thinking: 

What about ... darn. Not sure if there's a perfect word here. `python-from-PATH` would be most accurate but it's too wordy :/
What about simply `Path`?

**Comment by native-api** [comment]
> `system` could also be taken to mean the python bundled with the system

Sure it can -- but it's not the common meaning of the word, which is -- _"provided_ by the system".
That's because a system is not limited to what someone, at some point, "bundled" with it -- it, by design, allows you to install other software, too, and even replace the "bundled" stuff (e.g. with stuff installed from PPAs).

**Comment by aiguofer** [comment]
IMO `system` is pretty indicative of what it is.... if you didn't have `pyenv` installed and you just ran `python`, that's your "system python", and that's exactly what runs when `pyenv` is set to `system`. There's a million ways to configure python and it's pretty hard to account for all of them. You could install from a package manager, from homebrew, from scratch.... the idea with `pyenv` is that you do all that directly inside `pyenv`, and the only other python out there is the `system` python, w/e that may be.

Changing the name would be disruptive to users if they have set the local or global version to `system`. It would also potentially break extensions that depend on the `system` label for w/e reason.

**Comment by native-api** [comment]
@caleb15 Mind you, we're happy to hear any feedback from our users and look for ways to improve UX. It's just, as you can see, we're hard-pressed to find just _anything else_ that'd be more representative of its nature and behavior :slightly_smiling_face:
